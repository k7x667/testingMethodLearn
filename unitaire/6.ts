/**
 * Vous avez ici une classe implémentant la design pattern observer et une autre classe qui vous permet de générer des observers (ObserverImplementation n'est pas à tester, seulement Subject)
 * Vérifiez que la classe fonctionne comme prévu
 *
 * Attention pensez bien à reset la classe avant chaque test, ou pour différents groupes de tests (sinon les informations persistent dans la classe)
 *
 */

export interface Observer {
  update(data: any): void;
}

export class Subject {
  private observers: Observer[] = [];

  getObservers(): Observer[] {
    return this.observers;
  }

  addObserver(observer: Observer): void {
    this.observers.push(observer);
  }

  removeObserver(observer: Observer): void {
    const index = this.observers.indexOf(observer);
    if (index !== -1) {
      this.observers.splice(index, 1);
    }
  }

  notify(data: any): void {
    for (const observer of this.observers) {
      observer.update(data);
    }
  }
}

export class ObserverImplementation implements Observer {
  private name: string;

  constructor(name: string) {
    this.name = name;
  }

  update(data: any): void {
    console.log(`${this.name} received update with data: ${data}`);
  }
}
