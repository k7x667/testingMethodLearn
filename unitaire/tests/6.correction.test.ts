import { test, expect, vi, describe, beforeEach } from 'vitest';
import { Subject, Observer, ObserverImplementation } from '../6';

test('Test can add an observer', () => {
  const subject = new Subject();
  const observer = new ObserverImplementation('test');

  subject.addObserver(observer);
  const observers = subject.getObservers();

  expect(Array.isArray(observers)).toBeTruthy();
  expect(observers).toMatchSnapshot();
});

test('Test can add multiple observer', () => {
  const subject = new Subject();
  const observer = new ObserverImplementation('test');
  const observer2 = new ObserverImplementation('Test 2');

  subject.addObserver(observer);
  subject.addObserver(observer2);
  const observers = subject.getObservers();

  expect(Array.isArray(observers)).toBeTruthy();
  expect(observers).toMatchSnapshot();
});

test('Test can delete an observer when one', () => {
  const subject = new Subject();
  const observer = new ObserverImplementation('test');

  subject.addObserver(observer);
  subject.removeObserver(observer);
  const observers = subject.getObservers();

  expect(observers).toEqual([]);
});

test('Test can delete an observer when multiple', () => {
  const subject = new Subject();
  const observer = new ObserverImplementation('test');
  const observer2 = new ObserverImplementation('Test 2');

  subject.addObserver(observer);
  subject.addObserver(observer2);
  subject.removeObserver(observer);
  const observers = subject.getObservers();

  expect(observers).toMatchSnapshot();
});

test('Notify launch observer function', () => {
  const subject = new Subject();
  const observer = new ObserverImplementation('test');
  const spyObserver = vi.spyOn(observer, 'update');

  subject.addObserver(observer);
  subject.notify('hello world');

  expect(spyObserver).toHaveBeenCalledOnce();
});

test('Notify launch all observers function', () => {
  const subject = new Subject();
  const observer = new ObserverImplementation('test');
  const observer2 = new ObserverImplementation('tes2');

  const spyObserver = vi.spyOn(observer, 'update');
  const spyObserver2 = vi.spyOn(observer2, 'update');

  subject.addObserver(observer);
  subject.addObserver(observer2);
  subject.notify('hello world');

  expect(spyObserver).toHaveBeenCalledOnce();
  expect(spyObserver).toHaveBeenCalledWith('hello world');
  expect(spyObserver2).toHaveBeenCalledOnce();
  expect(spyObserver2).toHaveBeenCalledWith('hello world');
});
