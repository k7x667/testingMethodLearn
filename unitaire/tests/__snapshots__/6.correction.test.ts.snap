// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test can add an observer 1`] = `
[
  ObserverImplementation {
    "name": "test",
  },
]
`;

exports[`Test can add multiple observer 1`] = `
[
  ObserverImplementation {
    "name": "test",
  },
  ObserverImplementation {
    "name": "Test 2",
  },
]
`;

exports[`Test can delete an observer when multiple 1`] = `
[
  ObserverImplementation {
    "name": "Test 2",
  },
]
`;
